### PR TITLE
Refactoring & Performance Improvements

### PR DESCRIPTION
**Pull request by wdanilo**
### Pull Request Description
- Implements [Performance Debugging](https://github.com/luna/ide/issues/230). The scene runs now 60 FPS on full-screen without any framerate drop-downs. The rendering routine was simplified drastically and display object management was re-written from scratch in a much more optimal way.
- Implements [WebGL performance monitoring](https://github.com/luna/ide/issues/77) which was needed in order to optimize performance. Now the performance monitoring tool is handled in an optimal way (it consumes 0 resources if not used). Press the backtick button (``` ` ```) to enable it.
- Implements [Can't run the build artifact](https://github.com/luna/ide/issues/257) which was an accidental bug.
- Fixes [Option `--version` is returning 1.0.0 value for each build.](https://github.com/luna/ide/issues/236)

### Important Notes
- Not much new code. Mostly refactoring and optimizing the way data is being passed.
- Display object implementation uses much fewer callbacks now.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/core/src/display/render/passes/screen.rs [hunk 1]
```diff
@@ -23,12 +23,15 @@ impl ScreenRenderPass {
...
     pub fn new(world:&World) -> Self {
         let screen = Screen::new(world);
+        screen.hide();
         Self {screen}
     }
 }
 
 impl RenderPass for ScreenRenderPass {
     fn run(&mut self, _:&Context, _:&UniformScope) {
+        self.screen.show();
```
Comment (farmaazon) [inline lib/core/src/display/render/passes/screen.rs]: I don't get it. Why we hide the show screen for rendering and then hide? A comment would be helpful.

File: app/package.json [hunk 2]
```diff
@@ -1,5 +1,7 @@
 {
   "name": "root",
+  "version": "1.0.0",
+  "buildVersion": "42c4859",
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply app/package.json]: Done. The `buildVersion` field is now generated in a special file in the `generated` directory which is not tracked by git.

File: build/run.js [hunk 3]
```diff
@@ -199,12 +200,25 @@ async function dist_js () {
...
+    let commitHashCmd = await cmd.run_read('git',['rev-parse','--short','HEAD'])
+    let commitHash    = commitHashCmd.trim()
+    let configFile    = await fs.readFile(configPath)
+    let config        = JSON.parse(configFile)
+    if (config.buildVersion != commitHash) {
+        config.buildVersion = commitHash
+        await fs.writeFile(configPath,JSON.stringify(config,undefined,2))
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply build/run.js]: done as explained above

File: build/test.sh [hunk 4]
```diff
@@ -1,3 +1,2 @@
 #!/bin/bash
-cargo test &&
-cargo run --manifest-path=script/rust/Cargo.toml --bin test-all -- --headless --chrome
+cargo test && cargo run --manifest-path=build/rust/Cargo.toml --bin test-all -- --headless --chrome
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply build/test.sh]: file removed

File: lib/core/src/debug/monitor.rs [hunk 5]
```diff
@@ -122,87 +132,161 @@ impl Config {
...
+    pub fn new() -> Self { default() }
+}
+
+impl DomData {
+    /// Constructor.
+    pub fn new() -> Self { default() }
+}
+
+impl Default for DomData {
+    fn default() -> Self {
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/debug/monitor.rs]: Right, fixed.

File: lib/core/src/debug/monitor.rs [hunk 6]
```diff
@@ -122,87 +132,161 @@ impl Config {
...
-    context       : CanvasRenderingContext2d,
-    is_first_draw : bool,
+    user_config : Config,
+    config      : SamplerConfig,
+    width       : f64,
+    height      : f64,
+    dom         : Option<Dom>,
+    panels      : Vec<Panel>,
+    first_draw  : bool,
+    visible     : bool,
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/debug/monitor.rs]: That is such a good observation. Thanks for catching it!

File: lib/core/src/debug/monitor.rs [hunk 7]
```diff
@@ -215,64 +299,69 @@ impl Monitor {
...
+            let ratio = web::window().device_pixel_ratio();
+            let width = self.config.labels_width
+                + self.config.results_width
+                + self.config.plots_width
+                + 4.0 * self.config.margin
+                + self.config.outer_margin; // no outer_margin on the left side.
+            let mut height = self.config.outer_margin;
+            for _panel in &self.panels {
+                height += self.config.margin + self.config.panel_height;
+            }
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/debug/monitor.rs]: Changing as described via phone

File: lib/core/src/display/object.rs [hunk 8]
```diff
@@ -36,106 +52,62 @@ impl ParentBind {
...
-        let child = child.display_object();
-        self.children.insert(child)
-    }
-
-    pub fn remove_child_by_index(&mut self, index:usize) {
-        let opt_child = self.children.remove(index);
-        opt_child.for_each(|t| t.raw_unset_parent());
+impl Debug for Callbacks {
+    fn fmt(&self, f:&mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f,"Callbacks")
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/object.rs]: They woudl be printed like this when deriving. Leaving as is.

File: lib/core/src/display/object.rs [hunk 9]
```diff
@@ -145,35 +117,84 @@ fn fn_on_change(dirty:ChildDirty, ix:usize) -> OnChange { || dirty.set(ix) }
...
+        let transform        = CachedTransform :: new(logger.sub("transform")        , None);
+        let child_dirty      = ChildDirty      :: new(logger.sub("child_dirty")      , None);
+        let removed_children = RemovedChildren :: new(logger.sub("removed_children") , None);
+        let new_parent_dirty = NewParentDirty  :: new(logger.sub("new_parent_dirty") , ());
+        let visible          = true;
+        let callbacks        = default();
+        Self {logger,parent_bind,children,removed_children,event_dispatcher,transform,child_dirty,new_parent_dirty,visible,callbacks}
+    }
+
+    pub fn is_visible(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/object.rs]: It is in top-level. And it was earlier as well. This file needs to be docummented.

File: lib/core/src/display/object.rs [hunk 10]
```diff
@@ -306,84 +369,71 @@ impl DisplayObjectDataMut {
...
-    pub fn add_child<T:DisplayObject>(&self, child:T) {
+    /// Adds a new `Object` as a child to the current one.
+    pub fn add_child<T:Object>(&self, child:T) {
         self.clone_ref().add_child_take(child);
     }
 
-    /// Adds a new `DisplayObject` as a child to the current one.
-    pub fn add_child_take<T:DisplayObject>(self, child:T) {
+    /// Adds a new `Object` as a child to the current one.
+    pub fn add_child_take<T:Object>(self, child:T) {
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/object.rs]: self is passed as value. Adding docs.

File: lib/core/src/display/render/passes/display_object.rs [hunk 11]
```diff
@@ -38,6 +38,6 @@ impl RenderPass for DisplayObjectRenderPass {
         let arr2 = vec![0,0,0,0];
         context.clear_bufferfv_with_f32_array(Context::COLOR,0,&arr);
         context.clear_bufferuiv_with_u32_array(Context::COLOR,1,&arr2);
-        self.target.render();
+        self.target.render2();
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/core/src/display/render/passes/display_object.rs]: oops

File: lib/gui/src/shapes.rs [hunk 12]
```diff
@@ -219,49 +214,49 @@ pub fn on_frame
...
+//    handle.forget();
+//
+//    let target = mouse.on_up.event.clone_ref();
+//    let handle = mouse_manager.on_up.add(move |event:&OnUp| {
+//        target.emit(());
+//    });
+//    handle.forget();
+//
+//    mouse_manager
+//}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/gui/src/shapes.rs]: This file is described in the header as a big work in progress.

File: lib/system/web/src/lib.rs [hunk 13]
```diff
@@ -99,7 +70,7 @@ impl Drop for IgnoreContextMenuHandle {
 }
 
 /// Ignores context menu when clicking with the right mouse button.
-pub fn ignore_context_menu(target:&EventTarget) -> Result<IgnoreContextMenuHandle> {
+pub fn ignore_context_menu(target:&EventTarget) -> Option<IgnoreContextMenuHandle> {
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply lib/system/web/src/lib.rs]: Fixed as agreed via phone

File: lib/core/src/display/render/passes/screen.rs (hunk 1 above)
Comment (wdanilo) [reply lib/core/src/display/render/passes/screen.rs]: Added docs.

**Comment by wdanilo** [comment]
I'm merging it as we have confirmed it is OK via phone today.

**Pull request merged**